// html2canvas-pro@2.0.2/dist/lib/dom/node-type-guards.js
"use strict";
/**
 * DOM Node Type Guards
 * Extracted to break circular dependencies
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCustomElement = exports.isOLElement = exports.isLIElement = exports.isHTMLElementNode = exports.isSVGElementNode = exports.isTextNode = exports.isElementNode = void 0;
/**
 * Check if node is an Element
 */
const isElementNode = (node) => node.nodeType === Node.ELEMENT_NODE;
exports.isElementNode = isElementNode;
/**
 * Check if node is a Text node
 */
const isTextNode = (node) => node.nodeType === Node.TEXT_NODE;
exports.isTextNode = isTextNode;
/**
 * Check if element is an SVG element
 */
const isSVGElementNode = (element) => typeof element.className === 'object';
exports.isSVGElementNode = isSVGElementNode;
/**
 * Check if node is an HTML element
 */
const isHTMLElementNode = (node) => (0, exports.isElementNode)(node) && typeof node.style !== 'undefined' && !(0, exports.isSVGElementNode)(node);
exports.isHTMLElementNode = isHTMLElementNode;
/**
 * Check if node is an LI element
 */
const isLIElement = (node) => node.tagName === 'LI';
exports.isLIElement = isLIElement;
/**
 * Check if node is an OL element
 */
const isOLElement = (node) => node.tagName === 'OL';
exports.isOLElement = isOLElement;
/**
 * Check if element is a custom element
 * Custom elements must have a hyphen and cannot be SVG elements
 */
const isCustomElement = (element) => !(0, exports.isSVGElementNode)(element) && element.tagName.indexOf('-') > 0;
exports.isCustomElement = isCustomElement;
