// html2canvas-pro@2.0.2/dist/lib/render/canvas/__tests__/background-renderer.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("assert");
const background_renderer_1 = require("../background-renderer");
const context_1 = require("../../../core/context");
const bounds_1 = require("../../../css/layout/bounds");
const config_1 = require("../../../config");
describe('BackgroundRenderer', () => {
    it('should be instantiated', () => {
        const mockWindow = {
            document: {
                createElement: (_name) => {
                    let _href = '';
                    return {
                        set href(value) {
                            _href = value;
                        },
                        get href() {
                            return _href;
                        },
                        get protocol() {
                            return 'http:';
                        },
                        get hostname() {
                            return 'localhost';
                        },
                        get port() {
                            return '';
                        }
                    };
                }
            },
            location: { href: 'http://localhost/' }
        };
        const config = new config_1.Html2CanvasConfig({ window: mockWindow });
        const context = new context_1.Context({
            logging: false,
            imageTimeout: 15000,
            useCORS: false,
            allowTaint: false
        }, new bounds_1.Bounds(0, 0, 800, 600), config);
        const canvas = {
            width: 800,
            height: 600
        };
        const ctx = {
            fillStyle: '',
            save: () => { },
            restore: () => { }
        };
        const deps = {
            ctx,
            context,
            canvas,
            options: {
                width: 800,
                height: 600,
                scale: 1
            }
        };
        const renderer = new background_renderer_1.BackgroundRenderer(deps);
        (0, assert_1.ok)(renderer);
    });
});
