// html2canvas-pro@2.0.2/dist/lib/render/canvas/__tests__/effects-renderer.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("assert");
const effects_renderer_1 = require("../effects-renderer");
describe('EffectsRenderer', () => {
    it('should be instantiated', () => {
        const ctx = {
            shadowOffsetX: 0,
            shadowOffsetY: 0,
            shadowBlur: 0,
            shadowColor: '',
            globalAlpha: 1,
            save: () => { },
            restore: () => { }
        };
        const deps = {
            ctx
        };
        const pathCallback = {
            path: () => { }
        };
        const renderer = new effects_renderer_1.EffectsRenderer(deps, pathCallback);
        (0, assert_1.ok)(renderer);
        // Test public methods exist
        (0, assert_1.strictEqual)(typeof renderer.applyEffects, 'function');
        (0, assert_1.strictEqual)(typeof renderer.applyEffect, 'function');
        (0, assert_1.strictEqual)(typeof renderer.popEffect, 'function');
    });
});
